src/java.base/share/classes/sun/security/ssl/SSLSocketInputRecord.java

```diff
@@ -242,6 +262,8 @@ private Plaintext[] decodeInputRecord() throws IOException, BadPaddingException
         }
         readFully(contentLen);
         recordBody.flip();
+        // Record is ready to be decoded, save it.
+        saveLastDecodeRecord();
```

Comment:
It might be heavy to cache last decode record for every input record.

Reply:
Yes, but only if you consider the low probability of needing this record. Overall performance impact should be negligible considering all the other operations we do. I couldn't think of a better way of passing this record upstream, unless we restrict `saveLastDecodeRecord` to contentLen of `2` which will make this not a general purpose method.

Reply:
The last record could be huge and keep in the memory for a while.  It may be not required to cache it if we are able to close the connection while receiving an alert.  For TLS 1.3, the connection should be closed for alter message.